File: android/src/main/kotlin/io/falconmetrics/falconmetrics_flutter/ConvertProtobuf.kt
package io.falconmetrics.falconmetrics_flutter

import io.falconmetrics.sdk.AddedToCartEvent
import io.falconmetrics.sdk.CompleteRegistrationEvent
import io.falconmetrics.sdk.CustomEvent
import io.falconmetrics.sdk.PurchaseEvent
import io.falconmetrics.sdk.SubscribeEvent
import io.falconmetrics.sdk.TrackingEvent
import pb.Event


fun convertTrackingEvent(event: Event.TrackingEvent): TrackingEvent {
    return when (event.eventCase) {
        Event.TrackingEvent.EventCase.COMPLETE_REGISTRATION -> {
            CompleteRegistrationEvent()
        }

        Event.TrackingEvent.EventCase.ADDED_TO_CART -> {
            AddedToCartEvent(
                itemId = event.addedToCart.itemId,
                quantity = event.addedToCart.quantity,
                productPriceInCents = event.addedToCart.productPriceInCents,
                currency = event.addedToCart.currency,
                productCategory = event.addedToCart.productCategory,
                cartId = event.addedToCart.cartId
            )
        }

        Event.TrackingEvent.EventCase.SUBSCRIBE -> {
            SubscribeEvent(
                currency = event.subscribe.currency,
                predictedLtvValueInCents = event.subscribe.predictedLtvValueInCents
            )
        }

        Event.TrackingEvent.EventCase.PURCHASE -> {
            PurchaseEvent(
                itemIds = event.purchase.itemIdsList,
                quantity = event.purchase.quantity,
                transactionId = event.purchase.transactionId,
                productPriceInCents = event.purchase.productPriceInCents,
                currency = event.purchase.currency,
                revenueInCents = event.purchase.revenueInCents,
                productCategory = event.purchase.productCategory,
                cartId = event.purchase.cartId,
                paymentMethod = event.purchase.paymentMethod,
                taxInCents = event.purchase.taxInCents,
                shippingCostInCents = event.purchase.shippingCostInCents,
                discountInCents = event.purchase.discountInCents
            )
        }

        Event.TrackingEvent.EventCase.CUSTOM_EVENT -> {

            val attributes = mutableMapOf<String, Any>()
            for ((key, value) in event.customEvent.attributesMap) {
                when (value.valueCase) {
                    Event.AttributeValue.ValueCase.STRING_VALUE -> {
                        attributes[key] = value.stringValue
                    }

                    Event.AttributeValue.ValueCase.INT_VALUE -> {
                        attributes[key] = value.intValue
                    }

                    Event.AttributeValue.ValueCase.DOUBLE_VALUE -> {
                        attributes[key] = value.doubleValue
                    }

                    Event.AttributeValue.ValueCase.BOOL_VALUE -> {
                        attributes[key] = value.boolValue
                    }

                    else -> {
                        //ignore
                    }
                }
            }

            CustomEvent(
                eventName = event.customEvent.eventName,
                attributes = attributes,
                currency = event.customEvent.currency,
                revenueInCents = event.customEvent.revenueInCents
            )
        }


        Event.TrackingEvent.EventCase.EVENT_NOT_SET -> {
            throw IllegalArgumentException("Event not set")
        }
    }

}

File: ios/falconmetrics_flutter/Sources/falconmetrics_flutter/ConvertProtobuf.swift
//
//  ConvertProtobuf.swift
//  falconmetrics_flutter
//
//  Created by FalconMetrics on 24/03/2025.
//
import FalconMetrics
import Foundation


func convertTrackingEvent(event: Pb_TrackingEvent) throws -> FalconMetrics.BaseEventBuilder {
    switch event.event {
    case .completeRegistration:
        return FalconMetricsSdk.shared.createCompleteRegistrationBuilder()

    case .addedToCart:
        return FalconMetricsSdk.shared.createAddToCartBuilder()
            .withItemId( event.addedToCart.itemID)
            .withQuantity(Int(event.addedToCart.quantity))
            .withProductPrice(Int(event.addedToCart.productPriceInCents))
            .withCurrency(event.addedToCart.currency)
            .withProductCategory( event.addedToCart.productCategory)
            .withCartId( event.addedToCart.cartID)

    case .subscribe:
        return FalconMetricsSdk.shared.createSubscriptionBuilder()
            .withCurrency( event.subscribe.currency)
            .withPredictedValueInCents(Int(event.subscribe.predictedLtvValueInCents))
        

    case .purchase:
        return FalconMetricsSdk.shared.createPurchaseBuilder()
            .withItemIds(event.purchase.itemIds)
            .withQuantity(Int(event.purchase.quantity))
            .withTransactionId( event.purchase.transactionID)
            .withProductPrice(Int(event.purchase.productPriceInCents))
            .withCurrency(event.purchase.currency)
            .withProductCategory(event.purchase.productCategory)
            .withCartId(event.purchase.cartID)
            .withPaymentMethod(event.purchase.paymentMethod)
            .withTax(Int( event.purchase.taxInCents))
            .withShipping(Int(event.purchase.shippingCostInCents))
            .withDiscount(Int(event.purchase.discountInCents))
        
    case .customEvent:
        return FalconMetricsSdk.shared.createCustomEventBuilder()
            .withAttributes(event.customEvent.attributes)
            .withEventName(event.customEvent.eventName)
            .withCurrency(event.customEvent.currency)
            .withRevenueInCents(Int(event.customEvent.revenueInCents))

    @unknown default:
        throw NSError(domain: "TrackingError", code: 2, userInfo: [NSLocalizedDescriptionKey: "Unknown event case"])
    }


func convertUserData(_ userData: Pb_UserData) -> UserData {
    var builder = FalconMetricsSdk.shared.createUserDataBuilder()
    
    if userData.hasEmail && !userData.email.isEmpty {
        builder = builder.withEmail(userData.email)
    }
    if userData.hasPhoneNumber && !userData.phoneNumber.isEmpty {
        builder = builder.withPhoneNumber(userData.phoneNumber)
    }
    if userData.hasFirstName && !userData.firstName.isEmpty {
        builder = builder.withFirstName(userData.firstName)
    }
    if userData.hasLastName && !userData.lastName.isEmpty {
        builder = builder.withLastName(userData.lastName)
    }
    if userData.hasDateOfBirth && !userData.dateOfBirth.isEmpty {
        builder = builder.withDateOfBirth(userData.dateOfBirth)
    }
    if userData.hasCity && !userData.city.isEmpty {
        builder = builder.withCity(userData.city)
    }
    if userData.hasState && !userData.state.isEmpty {
        builder = builder.withState(userData.state)
    }
    if userData.hasPostalCode && !userData.postalCode.isEmpty {
        builder = builder.withPostalCode(userData.postalCode)
    }
    if userData.hasCountry && !userData.country.isEmpty {
        builder = builder.withCountry(userData.country)
    }
    
    return builder.build()
}
}
